### ISSUE TITLE
Dynamic  contact-user in mod_sofia

### ISSUE THREAD

**Issue by mohammaddehnavi** [signalwire/freeswitch#739]
Hello
How can I set the dynamic variable (I want to change mod_sofia in contact header to my caller number) in mod_sofia? 
I test this 
        <param name="extension-in-contact" value="true"/>
but not working.

**Comment by nttranbao** [comment]
You may try using this variable to set contact user for outbound calls..
`sip_contact_user`

**Comment by cmcom-gb** [comment]
@briankwest @anthmFS could you please provide an example on how to set the contact user dynamically? I've tried different things in the gateway params list, or in the dialplan, but am unable to set the contact user to the caller id (which is obviously different for each call to a particular gateway).

@nttranbao sip_contact_user seems to be readonly (from dialplan), and setting the value of extension to sip_contact user, or ${sip_contact_user}, $${sip_contact_user} doesn't work either.

**Comment by nttranbao** [comment]
Ah, I see you are using gateway, then yes the variable won't have any effect.

But if you originate/bridge without using gateway (use fspath), then it will.

Something like....
`fs_cli -x "originate {codec_string=PCMU}sofia/caller/911@emergency.local.ca;fs_path=sip:10.120.27.4:5060 &playback(phrase:funny_prompts)"`

**Comment by cmcom-gb** [comment]
@nttranbao thanks for answering. I'm actually using the bridge application. Does this also work with bridge, or are there other ways to do it? Using the gateway is required, since I'm using the distributor for load balancing between multiple gateways of the same carrier.

**Comment by nttranbao** [comment]
Yeah, this is the same issue I got before, when using gateway mode. I found no way to set the contact user dynamically for outbound INVITE. 
I ended up using OpenSIPS to rewrite it, and later using fs_path instead of gateway.
Maybe a custom source code patch may help then...

**Comment by nttranbao** [comment]
Well, while waiting for feedback from FreeSWITCH devs, I just want to see if there is a way to do that, with what we currently have...

My understanding is, mod_distributor will return a text string, which will be appended to the "sofia/gateway/" portion.
So, in theory, we can do something like below and achieve the same result....

1. Define a special distributor list, where node name is an "host:port", representing a specific proxy.
```
<configuration name="distributor.conf" description="Distributor Configuration">
  <lists>
    <!-- every 10 calls to test you will get foo1 once and foo2 9 times...yes NINE TIMES! -->
    <!-- this is not the same as 100 with 10 and 90 that would do foo1 10 times in a row then foo2 90 times in a row -->
    <list name="proxies" total-weight="10">
      <node name="1.1.1.1:5060" weight="1"/>
      <node name="2.2.2.2:5060" weight="9"/>
    </list>
  </lists>
</configuration>
```

2. Use that in the originator/bridge, with fs_path...
`fs_cli -x "originate {codec_string=PCMU}sofia/profile-name/911@emergency.local.ca;fs_path=sip:${distributor(proxies)} &playback(phrase:funny_prompts)"`

Note: If you are using dead gateway detection/exclusion, then it's still doable... In this case where distributor likely returns an online gateway name, we can either use DNS to resolve to host:IP, or save this in a variable and use XML and manually map to the right IP address and port..

My 2 cents...
